src/const-safety.md

```diff
@@ -0,0 +1,171 @@
+far as run-time Rust is concerned, but are inherently unsupportable in the miri engine and hence
+not const safe (they fall in the first category of failures above).
+We call these operations *unconst*.  The purpose
+of the following section is to explain this in more detail, before proceeding
+with the main definitions.
+
+## Miri background
+
+A very simple example of an unconst operation is
+
+```rust
+static S:i32 = 0;
+const BAD:bool = (&S as *const i32 as usize) % 16 == 0;
+```
+
+The modulo operation here is not supported by the miri engine because evaluating
+it requires knowing the actual integer address of `S`.
+
+The way miri handles this is by treating pointer and integer values separately.
+The most primitive kind of value in miri is a `Scalar`, and a scalar is *either*
+a pointer (`Scalar::Ptr`) or a bunch of bits representing an integer
+(`Scalar::Bits`).  Every value of a variable of primitive type is stored as a
+`Scalar`.  In the code above, casting the pointer `&S` to `*const i32` and then
+to `usize` does not actually change the value -- we end up with a local variable
+of type `usize` whose value is a `Scalar::Ptr`.  This is not a problem in
+itself, but then executing `%` on this *pointer value* is unsupported.
+
+However, it does not seem appropriate to blame the `%` operation above for this
+failure. `%` on "normal" `usize` values (`Scalar::Bits`) is perfectly fine, just using it on
+values computed from pointers is an issue.  Essentially, `&i32 as *const i32 as
+usize` is a "safe" `usize` at run-time (meaning that applying safe operations to
+this `usize` cannot lead to misbehavior, following terminology [suggested here])
+-- but the same value is *not* "safe" at compile-time, because we can cause a
+const safety violation by applying a safe operation (namely, `%`).
+
+This is similar to how it is illegal in Rust to have a dangling reference, because you
+can safely dereference it, which would be undefined behavior. Even though the issue
+occurs at the time of the dereference, we treat the creation of a dangling reference
+as the unsound action, not the final dereference.
+
+[suggested here]: https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html
+
+## Const safety check on values
+
+The result of any const computation (`const`, `static`, promoteds) is subject to
+a "sanity check" which enforces const safety.  Const safety
+is defined as follows:
+
+* Integer and floating point types are const-safe if they are a `Scalar::Bits`.
+  This makes sure that we can run `%` and other operations without violating
+  const safety.  In particular, the value must *not* be uninitialized.
+* References are const-safe if they are `Scalar::Ptr` into allocated memory, and
+  the data stored there is const-safe.  (Technically, we would also like to
+  require `&mut` to be unique and the memory behind a `&` to not be mutable unless there is an
+  `UnsafeCell`, but it seems infeasible to check that.)  For wide pointers, the
```

Comment:
We should make it clear that, similar to run-time, having aliasing mutable reference at const-time is still wrong even if the sanity check does not catch that. Likely the header above the list should make it clear that the sanity check *approximates* const safety but may actually accept some const-unsafe values.

Reply:
Alternatively... I don't think we have such a list for normal run-time safety, I think. So this is probably a level of detail that we do not yet want to commit to. Maybe it is better to remove this here and leave the discussion more high-level -- and only keep this level of detail in our const-eval writeup.